test-framework/src/main/java/org/elasticsearch/test/disruption/NetworkPartition.java

```diff
@@ -179,11 +174,13 @@ public synchronized void startDisrupting() {
         activeDisruption = true;
         for (String node1 : nodesSideOne) {
             MockTransportService transportService1 = (MockTransportService) cluster.getInstance(TransportService.class, node1);
-            DiscoveryNode discoveryNode1 = discoveryNode(node1);
             for (String node2 : nodesSideTwo) {
-                DiscoveryNode discoveryNode2 = discoveryNode(node2);
                 MockTransportService transportService2 = (MockTransportService) cluster.getInstance(TransportService.class, node2);
-                applyDisruption(discoveryNode1, transportService1, discoveryNode2, transportService2);
+                for (TransportAddress address1 : transportService1.boundAddress().boundAddresses()) {
+                    for (TransportAddress address2 : transportService2.boundAddress().boundAddresses()) {
+                        applyDisruption(address1, transportService1, address2, transportService2);
```

Comment:
publish address should normally be one of the bound transport addresses. If not, we run into some trouble in LookupTestTransport which uses publish address to determine transport. How about (instead of opening a can of worms) adding a check that requires publish address to be part of bound addresses, and fail otherwise?